### PR TITLE
FEAT: Exposed the array variable's DotDu value via coupledArrayDotDu

### LINKED ISSUE #25833: Accessing a coupled array's du_dot_du
### Discussed in https://github.com/idaholab/moose/discussions/25832

<div type='discussions-op-text'>

<sup>Originally posted by **NateAM** October 24, 2023</sup>
I'm having some trouble accessing a coupled array's _du_dot_du variable. In my header I define

`myClass.h`

    ...
    const VariableValue & _dv_dot_dv;
    ...

and in the source file I define

`myClass.C`

    InputParameters myClass::validParams( ){
        ...
        params.addRequiredCoupledVariable( "v", "my docstring" );
        ...
    }

    myClass::myClass( const InputParameters & parameters )
        : myBaseClass( parameters ),
        ...
        _v( coupledArrayValue( "v" ) ),
        _v_dot( coupledArrayDot( "v" ) ),
        _dv_dot_dv( coupledDotDu( "v" ) ),
        ...{
    }

When I run this, I get the moose error:

    *** ERROR ***
    Variable 'v' is of a different C++ type than you tried to fetch it as.

I will work to implement `coupledArrayDotDu` into `Coupleable`.

### PR DESCRIPTION
**Pull request by NateAM**
## Discussion

The current implementation of array variable's duDotDu returns a reference to a VariableValue because the derivatives are the same for all of the variables (due to identical time integration scheme). Consequently, the implementation of coupledArrayDotDu also returns a reference to a VariableValue.

closes #25833

### CODE REVIEW

**Comment by lindsayad** [comment]
Hi @NateAM, thanks for your pull request! let me know if you have any questions during the process. Looks like you'll need to apply some clang formatting. Hopefully the message is clear, but let me know if you have any trouble with that

**Review by lindsayad — COMMENTED** [review]
It would be nice to add a test for this. If you cannot come up with one, then I probably can

**Comment by NateAM** [comment]
> It would be nice to add a test for this. If you cannot come up with one, then I probably can

I could make something though I _think_ I would need to implement a new kernel to do it (an `ArrayCoupledTimeDereivative`). Maybe that goes beyond the scope of the PR?

**Comment by YaqiWang** [comment]
I think `duDotDu` only depends on the time integration scheme, thus for all components of an array variable, they are the same. You are returning `VariableValue`, which seems contradict with what you said in the message?

**Comment by NateAM** [comment]
> I think `duDotDu` only depends on the time integration scheme, thus for all components of an array variable, they are the same. You are returning `VariableValue`, which seems contradict with what you said in the message?

I've edited the original message to make my meaning (hopefully!) clear.

**Comment by lindsayad** [comment]
I might say:

The current implementation of array variable's duDotDu returns a reference to a VariableValue because the derivatives are the same for all of the variables (due to identical time integration scheme). Consequently, the implementation of coupledArrayDotDu also returns a reference to a VariableValue.

**Comment by NateAM** [comment]
@lindsayad I added a new kernel (`ArrayCoupledTimeDerivative`) to test the additions proposed here. Hopefully this helps get this PR merged.

**Comment by lindsayad** [comment]
thanks @NateAM !

File: test/tests/kernels/array_coupled_time_derivative/tests [hunk 1]
```diff
@@ -0,0 +1,29 @@
+[Tests]
+  [./jac_1d]
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/tests:2]:
```suggestion
  [jac_1d]
```

File: test/tests/kernels/array_coupled_time_derivative/tests [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    type = 'PetscJacobianTester'
+    input = 'test_jacobian.i'
+    cli_args = 'Outputs/exodus=false Mesh/dim=1 Mesh/nx=3'
+    run_sim = True
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 1d grid'
+  [../]
+  [./jac_2d]
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/tests:11]:
```suggestion
  []
  [jac_2d]
```

File: test/tests/kernels/array_coupled_time_derivative/tests [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    type = 'PetscJacobianTester'
+    input = 'test_jacobian.i'
+    cli_args = 'Outputs/exodus=false Mesh/dim=2 Mesh/nx=3 Mesh/ny=3'
+    run_sim = True
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 2d grid'
+  [../]
+  [./jac_3d]
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/tests:20]:
```suggestion
  []
  [jac_3d]
```

File: test/tests/kernels/array_coupled_time_derivative/tests [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    type = 'PetscJacobianTester'
+    input = 'test_jacobian.i'
+    cli_args = 'Outputs/exodus=false Mesh/dim=3 Mesh/nx=3 Mesh/ny=3 Mesh/nz=3'
+    run_sim = True
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 3d grid'
+  [../]
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/tests:28]:
```suggestion
  []
```

File: test/tests/kernels/array_coupled_time_derivative/tests (hunk 1 above)
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/tests:2]: We will need a `design` and `issues` for this. This is why the documentation target is failing. If you need any help on how to add those, just let me know!

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: test/tests/kernels/array_coupled_time_derivative/tests (hunk 1 above)
Comment (NateAM) [reply test/tests/kernels/array_coupled_time_derivative/tests:2]: Done

File: test/tests/kernels/array_coupled_time_derivative/tests (hunk 2 above)
Comment (NateAM) [reply test/tests/kernels/array_coupled_time_derivative/tests:11]: Done

File: test/tests/kernels/array_coupled_time_derivative/tests (hunk 3 above)
Comment (NateAM) [reply test/tests/kernels/array_coupled_time_derivative/tests:20]: Done

File: test/tests/kernels/array_coupled_time_derivative/tests (hunk 4 above)
Comment (NateAM) [reply test/tests/kernels/array_coupled_time_derivative/tests:28]: Done

File: test/tests/kernels/array_coupled_time_derivative/tests (hunk 1 above)
Comment (NateAM) [reply test/tests/kernels/array_coupled_time_derivative/tests]: @lindsayad I think I've gotten the `issues` and `design` sorted out. I basically copied `CoupledTimeDerivative` word for word with some small changes. Let me know if that does what you need.

File: framework/doc/content/source/kernels/ArrayCoupledTimeDerivative.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
+# ArrayCoupledTimeDerivative
+
+## Description
+
+The `ArrayCoupledTimeDerivative` kernel is very similar to the
+[`CoupledTimeDerivative`](/CoupledTimeDerivative.md) kernel with the
+exception that it works for array variables rather than scalars.
+The strong form on the the domain $\Omega$ is
```
Comment (lindsayad) [inline framework/doc/content/source/kernels/ArrayCoupledTimeDerivative.md:8]:
```suggestion
The strong form on the domain $\Omega$ is
```

File: framework/doc/content/source/kernels/ArrayCoupledTimeDerivative.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+## Example Syntax
+
+The syntax is simple, taking its type (`ArrayCoupledTimeDerivative`), the variable
+to that the residual the `ArrayCoupledTimeDerivative` contributes, and the coupled
+variable `v` that the time derivative operator acts upon. Example syntax can be
+found in the kernel block below:
+
+!listing
+test/tests/kernels/array_coupled_time_derivative/test_jacobian.i
+block=Kernels label=false
```
Comment (lindsayad) [inline framework/doc/content/source/kernels/ArrayCoupledTimeDerivative.md:42]: This doesn't show up correctly in the documentation (nor does it for the `CoupledTimeDerivative.md` file). I think the issue might be the line breaks. Can you try removing the line breaks?

File: framework/include/interfaces/Coupleable.h [hunk 7]
```diff
@@ -1084,6 +1084,17 @@ class Coupleable
...
                                                 unsigned int comp = 0) const;
 
+  /**
+   * Time derivative of a coupled array variable with respect to the coefficients
+   * @param var_name Name of coupled vector variable
+   * @param comp Component number for vector of coupled vector variables
+   * @return Reference to a ArrayVariableValue containing the time derivative of the coupled
+   * variable
+   */
+
```
Comment (lindsayad) [inline framework/include/interfaces/Coupleable.h:1094]:
```suggestion
```

File: framework/include/interfaces/Coupleable.h [hunk 8]
```diff
@@ -1084,6 +1084,17 @@ class Coupleable
...
+  virtual const VariableValue & coupledArrayDotDu(const std::string & var_name,
```
Comment (lindsayad) [inline framework/include/interfaces/Coupleable.h:1095]:
```suggestion
  const VariableValue & coupledArrayDotDu(const std::string & var_name,
```

Comment (lindsayad) [reply framework/include/interfaces/Coupleable.h:1095]: I don't know why the other ones are marked `virtual`.

File: framework/include/kernels/ArrayCoupledTimeDerivative.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  ArrayCoupledTimeDerivative(const InputParameters & parameters);
+
+protected:
+  virtual void computeQpResidual(RealEigenVector & residual) override;
+  virtual RealEigenVector computeQpJacobian() override;
+  virtual RealEigenMatrix computeQpOffDiagJacobian(const MooseVariableFEBase & jvar) override;
+
+  const ArrayVariableValue & _v_dot;
+  const VariableValue & _dv_dot;
+  const unsigned int _v_var;
```
Comment (lindsayad) [inline framework/include/kernels/ArrayCoupledTimeDerivative.h:31]: Some short doxygen would be good

File: framework/src/interfaces/Coupleable.C [hunk 10]
```diff
@@ -1427,6 +1427,31 @@ Coupleable::coupledDotDotDu(const std::string & var_name, unsigned int comp) con
   }
 }
 
+const VariableValue &
+Coupleable::coupledArrayDotDu(const std::string & var_name, unsigned int comp) const
+{
+  const auto * var = getArrayVar(var_name, comp);
```
Comment (lindsayad) [inline framework/src/interfaces/Coupleable.C:1433]:
```suggestion
  const auto * const var = getArrayVar(var_name, comp);
```

File: framework/src/kernels/ArrayCoupledTimeDerivative.C [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+registerMooseObject("MooseApp", ArrayCoupledTimeDerivative);
+
+InputParameters
+ArrayCoupledTimeDerivative::validParams()
+{
+  InputParameters params = ArrayKernel::validParams();
+  params.addClassDescription(
+      "Time derivative Array Kernel that acts on a coupled variable. Weak form: "
+      "$(\\psi_i, \\frac{\\partial v_h}{\\partial t})$. The coupled varialbe and"
```
Comment (lindsayad) [inline framework/src/kernels/ArrayCoupledTimeDerivative.C:20]:
```suggestion
      "$(\\psi_i, \\frac{\\partial v_h}{\\partial t})$. The coupled variable and"
```

File: framework/src/kernels/ArrayCoupledTimeDerivative.C [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    _v_dot(coupledArrayDot("v")),
+    _dv_dot(coupledArrayDotDu("v")),
+    _v_var(coupled("v"))
+{
+}
+
+void
+ArrayCoupledTimeDerivative::computeQpResidual(RealEigenVector & residual)
+{
+
```
Comment (lindsayad) [inline framework/src/kernels/ArrayCoupledTimeDerivative.C:37]:
```suggestion
```

File: test/tests/kernels/array_coupled_time_derivative/test_jacobian.i [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  [v_time_derivative]
+    type = ArrayTimeDerivative
+    variable = v
+    use_displaced_mesh = false
+  []
+  [v_diffusion]
+    type = ArrayDiffusion
+    variable = v
+    diffusion_coefficient = v_dc
+    use_displaced_mesh = false
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/test_jacobian.i:43]:
```suggestion
  []
  [u_time_derivative]
    type = ArrayTimeDerivative
    variable = u
  []
  [u_diffusion]
    type = ArrayDiffusion
    variable = u
    diffusion_coefficient = u_dc
  []
  [v_time_derivative]
    type = ArrayTimeDerivative
    variable = v
  []
  [v_diffusion]
    type = ArrayDiffusion
    variable = v
    diffusion_coefficient = v_dc
```

File: test/tests/kernels/array_coupled_time_derivative/test_jacobian.i [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+  []
+  [v_dc]
+    type = GenericConstantArray
+    prop_name = v_dc
+    prop_value = '2 2'
+  []
+[]
+
+[Preconditioning]
+  [./smp]
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/test_jacobian.i:74]:
```suggestion
  [smp]
```

File: test/tests/kernels/array_coupled_time_derivative/test_jacobian.i [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    type = SMP
+    full = true
+  [../]
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/test_jacobian.i:77]:
```suggestion
  []
```

File: test/tests/kernels/array_coupled_time_derivative/tests [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  [jac_1d]
+    type = 'PetscJacobianTester'
+    input = 'test_jacobian.i'
+    cli_args = 'Outputs/exodus=false Mesh/dim=1 Mesh/nx=3'
+    run_sim = True
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    issues = '#25833'
+    design = 'kernels/ArrayCoupledTimeDerivative.md'
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 1d grid'
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/tests:11]:
```suggestion
    requirement = 'The system shall be able to reproduce the exact Jacobian for a 1d grid for a coupled array time derivative kernel.'
```

File: test/tests/kernels/array_coupled_time_derivative/tests [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  [jac_2d]
+    type = 'PetscJacobianTester'
+    input = 'test_jacobian.i'
+    cli_args = 'Outputs/exodus=false Mesh/dim=2 Mesh/nx=3 Mesh/ny=3'
+    run_sim = True
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    issues = '#25833'
+    design = 'kernels/ArrayCoupledTimeDerivative.md'
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 2d grid'
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/tests:22]:
```suggestion
    requirement = 'The system shall be able to reproduce the exact Jacobian for a 2d grid for a coupled array time derivative kernel.'
```

File: test/tests/kernels/array_coupled_time_derivative/tests [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+  [jac_3d]
+    type = 'PetscJacobianTester'
+    input = 'test_jacobian.i'
+    cli_args = 'Outputs/exodus=false Mesh/dim=3 Mesh/nx=3 Mesh/ny=3 Mesh/nz=3'
+    run_sim = True
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    issues = '#25833'
+    design = 'kernels/ArrayCoupledTimeDerivative.md'
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 3d grid'
```
Comment (lindsayad) [inline test/tests/kernels/array_coupled_time_derivative/tests:33]:
```suggestion
    requirement = 'The system shall be able to reproduce the exact Jacobian for a 3d grid for a coupled array time derivative kernel.'
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
great contribution! Just some small suggested changes

File: framework/doc/content/source/kernels/ArrayCoupledTimeDerivative.md (hunk 6 above)
Comment (NateAM) [reply framework/doc/content/source/kernels/ArrayCoupledTimeDerivative.md]: I've tried to mimic one that seems to work correctly.

File: framework/include/kernels/ArrayCoupledTimeDerivative.h (hunk 9 above)
Comment (NateAM) [reply framework/include/kernels/ArrayCoupledTimeDerivative.h]: Done

**Comment by lindsayad** [comment]
@NateAM FYI I just did a force push squashing some of those commits

**Comment by lindsayad** [comment]
you are correct

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
as stated earlier, failure is unrelated to this PR

**Pull request merged**